### PR TITLE
tdem inductive sources and permeable targets

### PR DESCRIPTION
**Pull request by lheagy**
# tl; dr
- inductive source bug fix for permeable targets
- solve the magnetostatic problem for inductive sources that have initial fields
- new waveforms: trapezoid, quarter-sine ramp on

# Bug Fix: TDEM Inductive sources + permeable targets
How we initially calculated the source term for inductive sources when the model includes permeability was incorrect (I found this by comparing a magnetostatic solution with a late on-time solution. The late on-time solution matched the case where no permeable material was included), this pr fixes that for the E-B formulation.
image

In addition, we now have the option to solve the magnetostatic problem for initial fields when permeable material is included, so a shut-off waveform can be used when permeable materials are in the domain 

There is now a test to ensure that inductive sources are treating permeability properly

# New Waveforms
- `TrapezoidWaveform` and `QuarterSineRampOnWaveform`

```python
offTime=10
quarter_sine = TDEM.Src.QuarterSineRampOnWaveform(ramp_on=np.r_[0., 3], ramp_off=offTime - np.r_[1., 0] )
trapezoid = TDEM.Src.TrapezoidWaveform(ramp_on=np.r_[0., 3], ramp_off=offTime - np.r_[1., 0] )

nT = 1000
time = 10.*np.arange(nT)/nT
quarter_sine_plt = np.r_[[quarter_sine.eval(t) for t in time]]
trapezoid_plt = np.r_[[trapezoid.eval(t) for t in time]]


fig, ax = plt.subplots(1, 2, figsize=(8, 4))
ax[0].plot(time, quarter_sine_plt)
ax[0].set_title('quarter sine waveform')

ax[1].plot(time, trapezoid_plt)
ax[1].set_title('trapezoid waveform')

plt.tight_layout()
```

image

### CODE REVIEW

File: SimPEG/EM/TDEM/SrcTDEM.py [hunk 1]
```diff
@@ -311,36 +427,29 @@ def s_e(self, prob, time):
         b = self._bSrc(prob)
 
         if prob._formulation == 'EB':
-
-            MfMui = prob.MfMui
+            MfMui = prob.mesh.getFaceInnerProduct(1./self.mu)
```
Comment (lheagy) [inline SimPEG/EM/TDEM/SrcTDEM.py]: @sgkang: this was the problem

**Comment by rowanc1** [comment]
Are those pretty pictures in the docs somewhere?

**Comment by lheagy** [comment]
not yet, just in the tests at the moment. I could add an example

**Comment by lheagy** [comment]
now there is an example: 
image

**Comment by lheagy** [comment]
There is now also an example showing the available TDEM waveforms
image

File: SimPEG/EM/Static/DC/IODC.py [hunk 2]
```diff
@@ -225,6 +222,12 @@ class IO(properties.HasProperties):
     n_spacing = None
     n_data = None
 
+    def __init__(self, **kwargs):
+        super(IO, self).__init__(**kwargs)
+        warnings.warn(
```
Comment (lheagy) [inline SimPEG/EM/Static/DC/IODC.py]: @rowanc1, @sgkang: I moved the IO warning inside the instantiation of DC IO rather than on import (it was throwing a warning on the import of SimPEG, which was a bit annoying...)

File: SimPEG/EM/TDEM/SrcTDEM.py [hunk 3]
```diff
@@ -277,6 +365,36 @@ def _aSrc(self, prob):
 
         return a
 
+    def _getAmagnetostatic(self, prob):
```
[the remark this answers is not in the document]
Comment (lheagy) [reply SimPEG/EM/TDEM/SrcTDEM.py]: For EM problems with a compact source and a compact target, I don't think the boundary conditions should make too much of an impact, but we can test against Neumann

File: SimPEG/EM/TDEM/SrcTDEM.py [hunk 4]
```diff
@@ -277,6 +365,36 @@ def _aSrc(self, prob):
...
+        if prob._formulation == 'EB':
```
[the remark this answers is not in the document]
Comment (lheagy) [reply SimPEG/EM/TDEM/SrcTDEM.py]: I can update the errors to include more explicit messages

File: SimPEG/EM/TDEM/SrcTDEM.py (hunk 1 above)
Comment (lheagy) [reply SimPEG/EM/TDEM/SrcTDEM.py]: `prob.MfMui` should take `mu`, but to calculate the source, we need to work with `mu_0` since this is what is used to calculate the vector potential `a`. Does that make sense?

File: SimPEG/EM/TDEM/SrcTDEM.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lheagy) [reply SimPEG/EM/TDEM/SrcTDEM.py]: I have started #680 to keep track of this
